zebra-chain/src/commitments/sapling.rs

```diff
@@ -0,0 +1,341 @@
+                .iter()
+                .enumerate()
+                .for_each(|(i, bit)| bits.set(i, *bit));
+
+            let mut tmp = jubjub::Fr::one();
+
+            if bits[0] {
+                tmp += &jubjub::Fr::one();
+            }
+
+            if bits[1] {
+                tmp += &jubjub::Fr::one().double();
+            }
+
+            if bits[2] {
+                tmp -= tmp.double();
+            }
+
+            if j > 0 {
+                // Inclusive range!
+                tmp *= (1..=(4 * j)).fold(jubjub::Fr::one(), |acc, _| acc.double());
+            }
+
+            m_i += tmp;
+        }
+
+        m_i
+    }
+
+    let mut result = jubjub::ExtendedPoint::identity();
+
+    // Split M into n segments of 3 * c bits, where c = 63, padding the last
+    // segment with zeros.
+    //
+    // https://zips.z.cash/protocol/protocol.pdf#concretepedersenhash
+    for (i, segment) in M.chunks(189).enumerate() {
+        result += I_i(domain, i as u32) * M_i(&segment);
+    }
+
+    result
+}
+
+/// Mixing Pedersen Hash Function
+///
+/// Used to compute ρ from a note commitment and its position in the note
+/// commitment tree.  It takes as input a Pedersen commitment P, and hashes it
+/// with another input x.
+///
+/// MixingPedersenHash(P, x) := P + [x]FindGroupHash^J^(r)(“Zcash_J_”, “”)
+///
+/// https://zips.z.cash/protocol/protocol.pdf#concretemixinghash
+#[allow(non_snake_case)]
+pub fn mixing_pedersen_hash(P: jubjub::ExtendedPoint, x: jubjub::Fr) -> jubjub::ExtendedPoint {
+    const J: [u8; 8] = *b"Zcash_J_";
+
+    P + find_group_hash(J, b"") * x
+}
+
+/// Construct a 'windowed' Pedersen commitment by reusing a Pederson hash
+/// construction, and adding a randomized point on the Jubjub curve.
+///
+/// WindowedPedersenCommit_r (s) := \
+///   PedersenHashToPoint(“Zcash_PH”, s) + [r]FindGroupHash^J^(r)(“Zcash_PH”, “r”)
+///
+/// https://zips.z.cash/protocol/protocol.pdf#concretewindowedcommit
+pub fn windowed_pedersen_commitment(r: jubjub::Fr, s: &BitVec<Lsb0, u8>) -> jubjub::ExtendedPoint {
+    const D: [u8; 8] = *b"Zcash_PH";
+
+    pedersen_hash_to_point(D, &s) + find_group_hash(D, b"r") * r
+}
+
+/// The randomness used in the Pedersen Hash for note commitment.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct CommitmentRandomness(jubjub::Fr);
+
+/// Note commitments for the output notes.
+#[derive(Clone, Copy, Deserialize, PartialEq, Serialize)]
+pub struct NoteCommitment(#[serde(with = "serde_helpers::AffinePoint")] pub jubjub::AffinePoint);
+
+impl fmt::Debug for NoteCommitment {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("NoteCommitment")
+            .field("u", &hex::encode(self.0.get_u().to_bytes()))
+            .field("v", &hex::encode(self.0.get_v().to_bytes()))
+            .finish()
+    }
+}
+
+impl From<[u8; 32]> for NoteCommitment {
+    fn from(bytes: [u8; 32]) -> Self {
+        Self(jubjub::AffinePoint::from_bytes(bytes).unwrap())
+    }
+}
```

Comment:
I don't think this (using `unwrap`) will work, because it can cause a panic if we use it on untrusted data.  There are two options:

1. Change this to a `TryFrom` and keep the `NoteCommitment` as storing an `AffinePoint`;
2. Keep this as a `From` and change the `NoteCommitment` to store bytes.

I'm not sure which is the best option, since it depends on how the `NoteCommitment` is used.  Point decompression is somewhat expensive, so ideally we'd avoid doing it unless we have to.  Do we expect to process a `NoteCommitment` every time it's created, or only sometimes (e.g., when validating a block)?  Is the `NoteCommitment` included in any structures included in a block? If so, we would be decompressing / compressing all of these points every time we received a block or loaded it from storage, even if we just wanted to hand it out to someone else.

Reply:
> I don't think this (using `unwrap`) will work, because it can cause a panic if we use it on untrusted data. There are two options:
> 
> 1. Change this to a `TryFrom` and keep the `NoteCommitment` as storing an `AffinePoint`;
> 2. Keep this as a `From` and change the `NoteCommitment` to store bytes.
> 
> I'm not sure which is the best option, since it depends on how the `NoteCommitment` is used. Point decompression is somewhat expensive, so ideally we'd avoid doing it unless we have to. Do we expect to process a `NoteCommitment` every time it's created, or only sometimes (e.g., when validating a block)? For instance, if the `NoteCommitment` were included in any structures included in a block, then we would be decompressing / compressing all of these points every time we received a block or loaded it from storage, even if we just wanted to hand it out to someone else. Alternatively, if we only create note commitments while working with them, then we always need the uncompressed representation, so there's no extra work.

So to serialize/deserialize a transaction with an `Output` description in `ShieldedData`, we need to parse and construct at least a scalar in `Fq`, which is supposed to be the u-coordinate of an `AffinePoint` corresponding to a `NoteCommitment`.  Similarly, if there are either `Spend`s or `Output`s, we need to serialize/deserialize the compressed `AffinePoint` corresponding to a `ValueCommitment`.

Reply:
Personally I would go with a `TryFrom`

Reply:
I rambled in the wrong place about why these wrap `AffinePoint`s so: https://github.com/ZcashFoundation/zebra/pull/596#discussion_r467364873

Reply:
So I did this and when calling inside `zcash_deserialize()`, I need to expose an `Error` type that can convert to or is `SerializationError`, so I think I'll call the `try_from` inside `impl Zcash(De)Serialize`

Reply:
> So I did this and when calling inside `zcash_deserialize()`, I need to expose an `Error` type that can convert to or is `SerializationError`, so I think I'll call the `try_from` inside `impl Zcash(De)Serialize`

This has been done and 'seems fine'.